usermods/photoresistor_sensor_mqtt_v2/photoresistor_sensor_mqtt_v2.cpp

```diff
@@ -0,0 +1,204 @@
+  int ldrPin = A0;
+  unsigned long lastTime = 0;
+  unsigned int lightValue = 0;
+  float lightPercentage = 0;
+  float lastPercentage = 0;
+  bool hassDiscoverySent = false;
+  bool ldrEnabled = true;
+  bool inverted = false;
+  bool initDone = false;
+  String topic_buf;
+
+public:
+  photoresistor_sensor_mqtt_v2() {}
+
+  void setup()
+  {
+    // register ldrPin
+    if ((ldrPin >= 0) && ldrEnabled)
+    {
+      if (!PinManager::allocatePin(ldrPin, false, PinOwner::UM_PHOTO_RESISTOR_MQTT))
+      {
+        ldrEnabled = false; // pin already in use -> disable usermod
+      }
+      else
+      {
+        pinMode(ldrPin, INPUT); // alloc success -> configure pin for input
+      }
+    }
+    else
+    {
+      ldrEnabled = false; // invalid pin -> disable usermod
+    }
+    initDone = true;
+  }
+
+  void connected()
+  {
+    if (WLED_MQTT_CONNECTED && HomeAssistantDiscovery)
+    {
+      publishDiscovery();
+    }
+  }
+
+  void loop()
+  {
+    if (initDone && ldrEnabled && WLED_MQTT_CONNECTED)
+    {
+      bool force_update = (millis() - lastTime) > (update_interval * 100); // to keep the mqtt alive if the light is not changing
+      bool regular_update = (millis() - lastTime) > update_interval;
+      if (regular_update || force_update)
+      {
+        lightValue = analogRead(ldrPin);
+#ifdef ESP32
+        const float mapping = 4096.0; // ESP32 ADC resolution is 12 bit, so 4096 values
+#else
+        const float mapping = 1024.0; // ESP8266 ADC resolution is 10 bit, so 1024 values
+#endif
+        if (inverted)
+        {
+          lightPercentage = 100.0 - ((float)lightValue * -1 + mapping) / (float)mapping * 100;
+        }
+        else
+        {
+          lightPercentage = ((float)lightValue * -1 + mapping) / (float)mapping * 100;
+        }
+
+        if (abs(lightPercentage - lastPercentage) > change_threshold || force_update)
+        {
+          if (WLED_MQTT_CONNECTED)
+          {
+            publishMqtt(lightPercentage);
+          }
+          lastTime = millis();
+          lastPercentage = lightPercentage;
+          // debug
+          char buf[64];
+          sprintf_P(buf, PSTR("LDR: %d (%d)"), (int)lightPercentage, lightValue);
+          DEBUG_PRINTLN(buf);
+        }
+      }
+    }
+  }
+
+  void addToConfig(JsonObject &root)
+  {
+    JsonObject top = root.createNestedObject(FPSTR(_name));
+    top["Enabled"] = ldrEnabled;
+    top["LDR Pin"] = ldrPin;
+    top["LDR update interval"] = update_interval;
+    top["LDR inverted"] = inverted;
+    top["change threshould"] = change_threshold;
+  }
+
+  bool readFromConfig(JsonObject &root)
+  {
+    int8_t oldLdrPin = ldrPin;
+    JsonObject top = root[FPSTR(_name)];
+    bool configComplete = !top.isNull();
+    configComplete &= getJsonValue(top["Enabled"], ldrEnabled);
+    configComplete &= getJsonValue(top["LDR Pin"], ldrPin);
+    configComplete &= getJsonValue(top["LDR update interval"], update_interval);
```

Comment:
These are `cfg.json` configuration item key names. They need not be descriptive. Use `addInfo()` to add description.